### ISSUE TITLE
Location in Dark Castle not recognized

### ISSUE THREAD

**Issue by rrooij** [rrooij/sd3save_editor#2]
A location in Dark Castle is not recognized. When opening the save, the ID can't be found in the list of locations.

Its ID is 17117. I've never seen a number this large and don't know the cause.

The save is attached with this issue. Rename the .txt extension to srm. The first entry in the save is the one with the problem.

[Seiken Densetsu 3 (Japan) [En by LNF+Neill Corlett+SoM2Freak v1.01].txt](https://github.com/rrooij/sd3save_editor/files/1868069/Seiken.Densetsu.3.Japan.En.by.LNF.Neill.Corlett.SoM2Freak.v1.01.txt)

**Comment by noptoboggan** [comment]
So this is strange, in the first save at `0x726` I indeed see `dd 42` which as you say is 17,117 and the game loads in the Dark Castle just fine with this value.

However, after loading that save I saved again without changing areas at the nearby statue and inspected the srm at `0x726` again, it was updated to `dd 02 = 733 =  Dark Castle, south`

**Comment by noptoboggan** [comment]
```
0x42 = 0100 0010
0x02 = 0000 0010
```
Possible unintended bit-flip? I suppose that doesn't explain how the game would load it fine.

**Comment by noptoboggan** [comment]
I had a theory that the game loads the location value as a 12 bit integer.

There are 1,113 location values, a 12 bit integer could hold 4,095.

```
dd 42 = 1101 1101 >0100< 0010
dd 02 = 1101 1101 >0000< 0010
```

Without the leading four bits from the most significant byte, both end up being 733. I thought maybe the game was only using 12 bits of data here, which could explain how both values were able to load the save in the same location just fine.

I modified the location type to use an adapter where I am truncating the leading four bits from the most significant byte, and returning the resulting int on decode; this allowed the app to successfully load the save from this issue with `dd 42` as `Dark Castle, south`.

If the game is indeed only using 12 bits from these 2 bytes of memory, it would mean that "any" values in the 4 bits that are being truncated should not affect where you load in a save. To test this, I made the encode function of the adapter add a number of different values for those 4 bits and what do you know, not just in the save file from this issue but also in a few of my personal saves, it didn't matter what values were in those 4 bits, the location you would load in to didn't change!

It seems that the encode function of the adapter can just have those 4 bits as `0`s (or `1`s, it doesn't appear to matter), while the decode function can mimic the game by using only 12 bits of data from those 2 bytes to form an integer which allows the app to properly see the save in this issue as in the Dark Castle.

The question remains as to why that bit was 1. Seeing as it was a single bit being 1 it's possible it may have been an unintended bit flip that occurred in game or during save, just in an area that didn't affect the game while loading.

Thoughts? If you would like to test this you may see the commit I made here dc681cd. I've included the adapter below for reference.

```python3
class LocationAdapter(Adapter):
    """Location data appears to be loaded as a 12bit integer by the game
       This works around the absence of such a type in python core, it functions
       by truncating 4 bits from the provided 16, and zeroes out the leading 4
       bits on writing
    """
    def _decode(self, obj, context):
        int16 = int.from_bytes(obj, byteorder='little')
        bit_str12 = bin(int16)[2:].zfill(16)[4:]
        return int(bit_str12, 2)

    def _encode(self, obj, context):
        bit_str = '0000%s' % bin(obj)[2:].zfill(16)[4:]
        return int(bit_str, 2).to_bytes(2, byteorder='little')
```

```python3
"location"/Int16sl,

# .. was changed to ..

"location"/LocationAdapter(Bytes(2)),
```

**Comment by rrooij** [comment]
I tested it on a few locations and it still works perfectly. Thanks a lot man. This could indeed explain the problem. Feel free to create a pull request.

I still don't know the cause of the "wrongly" saved file myself. It was my own save and I didn't do any  crazy stuff with it as far as I know.

By the way, I find it very nice that someone found this project haha. I thought that it was so niche that I would probably be the only contributor for years to come.

**Comment by noptoboggan** [comment]
Good stuff!

It's possible some small corruption happened in the emulator somewhere and ended up in the current areas location value or similar which got saved. I wasn't able to reproduce it, I suppose time will tell.

Haha, glad I found the project. I had played through the game a number of times before, and recently wanted to avoid some of the late-game grind. Went about editing the save ram manually. I ended up looking for an editor online after doing the checksum through the CLI a couple times as it was quite cumbersome. So, thanks!
